Fix #1202 by changing enchantments to ingredients

naqaden opened the pull request:
Thanks to @HenryLoenwind for the [suggestion](https://github.com/mezz/JustEnoughItems/issues/1202#issuecomment-410721993).

naqaden committed:
Change enchantments to ingredients

Relieves enchantments from relying on IDs.

naqaden committed:
Hide ID-based enchanted books

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/mezz/JustEnoughItems?pullRequest=1402) <br/>All committers have signed the CLA.

mezz committed:
Fix #1405 Center search bar not centered after toggling overlay

mezz commented on src/main/java/mezz/jei/plugins/vanilla/ingredients/EnchantDataRenderer.java:
```diff
@@ -0,0 +1,83 @@
+import net.minecraft.item.ItemStack;
+import net.minecraft.util.text.TextFormatting;
+
+public class EnchantDataRenderer implements IIngredientRenderer<EnchantmentData> {
```
You should pass in an instance of `IIngredientRenderer<ItemStack>` to the constructor of this renderer, and use it to render the enchanted books.

mezz reviewed (changes requested):
Thanks for contributing!
This looks good but there are a couple small things to fix before I can merge it.
1. We will need a way to bridge the gap with recipes that use enchanted books right now. I have something hacked into recipe lookups for items that hold fluids, we can add another case to support old recipes that use enchanted books there. See `mezz.jei.recipes.RecipeRegistry#getRecipeWrappers` for the fluidstack special case handling.
2. JEI's existing enchanted book anvil recipes need to be updated

naqaden committed:
Use ItemStackRenderer for EnchantDataRenderer

and minor clean up with 'enchant' abbreviation consistency and ItemEnchantedBook's existing enchanted book function.

naqaden committed:
Use ItemEnchantedBook#getEnchantedItemStack

naqaden committed:
Added special cases for EnchantmentData

naqaden commented:
Thanks for that guidance. I don't yet see where to update anvil recipes, so right now each enchanted book shows anvil recipes for all enchanted books.

There's also a small issue where the previously blank enchanted book on the JEI items screen now renders with Protection I on it.

HenryLoenwind committed:
Add a Bookmarks list to the left side of the screen (#1408)

mezz committed:
Minor cleanup

mezz committed:
Fixes for cleanup

Commit:
update(pt_BR) add bookmark strings (#1416)

iPopstop committed:
Update ru_ru.lang (Bookmarks) (#1418)

mezz committed:
Add API to allow utility mods to set areas where JEI should not draw

mezz committed:
Fixes for live-updating gui exclusion areas

mezz committed:
Fix handling mouse input in config guis

CrafterKina committed:
Update ja_jp.lang (#1428)

Regnander committed:
Update sv_se.lang (#1427)

Greg-21 committed:
Updated Polish translation (#1421)

mezz commented:
I am working on adding oredict as an ingredient as well.
I think this PR will be a bit easier after that change is in, could take a while longer to get everything working right though.

mezz committed:
minor cleanup

tterrag1098 committed:
Add support for anvil recipe left side being a list (#1419)

HenryLoenwind commented:
oredict may be a bit trickier, as there are instances were we'd have lists of oredicts and itemstacks. Would a way to lookup all ingredients' recipes of a cycling slot at once be better? (This may need a way for teh recipe to tell you if it is cycling ingredient alternatives or recipe variants.)

mezz commented:
I've thought a lot about that question, and I was thinking oredict is the right way forward because of tags in 1.13.
With tags, it's even easier to create small groups of items for recipes.

That said, I'm not really convinced it's the right way either. I have implemented it in a branch so I can play around with it and see what corner cases I run into. So far there are a few pretty big ones where mods don't use oredict but still have large lists of items.

I am going to break off some pieces of my oredict branch that might benefit this PR. 
After that I will try the other approach, which is improving the way cycled ingredients are displayed and  navigated.

mezz committed:
Organize ingredients into subfolders

mezz committed:
Move some specially-handled ItemStack cases into the vanilla plugin

mezz commented:
This PR is pretty close to done, so I will update it and merge it in.

mezz committed:
Merge branch '1.12' of https://github.com/naqaden/JustEnoughItems into naqaden-1.12

mezz commented:
I ended up changing this pretty dramatically.
Now the "enchant" ingredient is pretty much hidden from mods and the players, it will just act like the regular enchanted books.
I tested it in a few packs and situations where mods are added or removed, and it seems to work as intended now. Mods will not have to add support, it just fixes the ingredient list issue and everything worked out.
